File: Zig/problem017.zig
// zig version 0.13.0
const std = @import("std");
const stdout = std.io.getStdOut().writer();

// #17 Number Letter Counts - Project Euler
// http://projecteuler.net/problem=17

fn count_strings(str: []const u8) usize {
    var count: usize = 0;
    var it = std.mem.splitScalar(u8, str, ' ');
    while (it.next()) |seq| {
        count += seq.len;
    }
    return count;
}

fn count_from_1_to_9() usize {
    return count_strings("one two three four five six seven eight nine");
}

fn count_from_10_to_19() usize {
    return count_strings("ten eleven twelve thirteen fourteen fifteen sixteen seventeen eighteen nineteen");
}

fn count_from_20_to_99() usize {
    const tys = count_strings("twenty thirty forty fifty sixty seventy eighty ninety");
    return tys * 10 + count_from_1_to_9() * 8;
}

fn count_from_1_to_99() usize {
    return count_from_1_to_9() + count_from_10_to_19() + count_from_20_to_99();
}

fn count_from_100_to_999() usize {
    return count_from_1_to_9() * 100 + "hundred".len * 900 + "and".len * (900 - 9) + count_from_1_to_99() * 9;
}

fn count_1000() usize {
    return count_strings("one thousand");
}

pub fn main() !void {
    const ans = count_from_1_to_99() + count_from_100_to_999() + count_1000();

    try stdout.print("{}\n", .{ans});
}

File: Kotlin/problem017.kt
// #17 Number Letter Counts - Project Euler
// http://projecteuler.net/problem=17

fun countStrings(string: String): Int {
    return string.split(" ").map { it.length }.sum()
}

fun countFrom1To9(): Int {
    return countStrings("one two three four five six seven eight nine")
}

fun countFrom10To19(): Int {
    return countStrings("ten eleven twelve thirteen fourteen fifteen sixteen seventeen eighteen nineteen")
}

fun countFrom20To99(): Int {
    return "twenty thirty forty fifty sixty seventy eighty ninety".split(" ").map { it.length * 10 + countFrom1To9()}.sum()
}

fun countFrom1To99(): Int {
    return countFrom1To9()  + countFrom10To19() + countFrom20To99()
}

fun countFrom100To999(): Int {
    return countFrom1To9() * 100 + "hundred".length * 900  + "and".length * (900 - 9)     + countFrom1To99() * 9
}

fun count1000(): Int {
    return countStrings("one thousand")
}

fun main() {
    val ans = countFrom1To99() + countFrom100To999() + count1000()
    println(ans)
}
